File: backend/app/api/v1/endpoints/validation.py
"""API endpoints for validation and dependency analysis."""

from typing import Any

from fastapi import APIRouter
from loguru import logger

from backend.app.models.fix import FixResult, FixSuggestion
from backend.app.models.project import Project
from backend.app.models.validation import DataValidationMode, ValidationError, ValidationResult
from backend.app.services.auto_fix_service import AutoFixService
from backend.app.services.dependency_service import DependencyGraph, DependencyService, get_dependency_service
from backend.app.services.project_service import ProjectService, get_project_service
from backend.app.services.validation_service import ValidationService, get_validation_service
from backend.app.utils.error_handlers import handle_endpoint_errors

router = APIRouter()


@router.post("/projects/{name}/validate/data", response_model=ValidationResult)
@handle_endpoint_errors
async def validate_project_data(
    name: str,
    entity_names: list[str] | None = None,
    validation_mode: DataValidationMode = DataValidationMode.SAMPLE,
) -> ValidationResult:
    """
    Run data-aware validation on project.

    This validates actual data (not just project structure):
    - Checks configured columns exist in data
    - Verifies natural keys are unique
    - Confirms entities return data
    - Validates foreign key data integrity

    Args:
        name: Project name
        entity_names: Optional list of entity names to validate (None = all entities)
        validation_mode: Validation mode:
            - SAMPLE: Fast validation using preview samples (up to 1000 rows per entity)
            - COMPLETE: Full validation using complete normalized dataset (slower but comprehensive)

    Returns:
        Validation result with data-specific errors and warnings
    """
    validation_service = get_validation_service()
    result = await validation_service.validate_project_data(
        name,
        entity_names,
        validation_mode=validation_mode,
    )

    logger.info(
        f"Data validation for '{name}' completed (mode={validation_mode.value}): "
        f"valid={result.is_valid}, errors={result.error_count}, warnings={result.warning_count}"
    )
    return result


@router.post("/projects/{name}/entities/{entity_name}/validate", response_model=ValidationResult)
@handle_endpoint_errors
async def validate_entity(name: str, entity_name: str) -> ValidationResult:
    """
    Validate specific entity within a project (structural validation only).

    Args:
        name: Project name
        entity_name: Entity name to validate

    Returns:
        Validation result with entity-specific errors and warnings
    """

    project_service: ProjectService = get_project_service()
    validation_service: ValidationService = get_validation_service()

    project: Project = project_service.load_project(name)
    result: ValidationResult = validation_service.validate_entity(project, entity_name)

    logger.info(f"Validated entity '{entity_name}' in project '{name}': " f"valid={result.is_valid}, errors={result.error_count}")

    return result


@router.get("/projects/{name}/dependencies")
@handle_endpoint_errors
async def get_dependencies(name: str) -> dict[str, Any]:
    """
    Get dependency graph for configuration.

    Analyzes entity dependencies including:
    - Source dependencies
    - Explicit depends_on declarations
    - Foreign key relationships

    Args:
        name: Project name

    Returns:
        Dependency graph with nodes, edges, cycles, and topological order
    """
    project_service: ProjectService = get_project_service()
    dependency_service: DependencyService = get_dependency_service()

    project: Project = project_service.load_project(name)
    graph: DependencyGraph = dependency_service.analyze_dependencies(project)
    logger.debug(f"Retrieved dependency graph for '{name}': " f"{len(graph['nodes'])} nodes, {len(graph['edges'])} edges")

    return dict(graph)


@router.post("/projects/{name}/dependencies/check")
@handle_endpoint_errors
async def check_dependencies(name: str) -> dict[str, Any]:
    """
    Check for circular dependencies in project.

    Args:
        name: Project name

    Returns:
        Dictionary with has_cycles flag, list of cycles, and cycle count
    """
    project_service: ProjectService = get_project_service()
    dependency_service: DependencyService = get_dependency_service()

    project: Project = project_service.load_project(name)
    result = dependency_service.check_circular_dependencies(project)

    logger.info(f"Checked circular dependencies for '{name}': " f"has_cycles={result['has_cycles']}, cycles={result['cycle_count']}")

    return result


@router.post("/projects/{name}/fixes/preview")
@handle_endpoint_errors
async def preview_fixes(name: str, errors: list[dict[str, Any]]) -> dict[str, Any]:
    """
    Preview automatic fixes for validation errors.

    Args:
        name: Project name
        errors: List of validation errors to generate fixes for

    Returns:
        Preview of fixes that would be applied
    """

    project_service: ProjectService = get_project_service()
    auto_fix_service = AutoFixService(project_service)

    validation_errors: list[ValidationError] = [ValidationError(**error) for error in errors]
    suggestions: list[FixSuggestion] = auto_fix_service.generate_fix_suggestions(validation_errors)
    preview: dict[str, Any] = await auto_fix_service.preview_fixes(name, suggestions)

    return preview


@router.post("/projects/{name}/fixes/apply")
@handle_endpoint_errors
async def apply_fixes(name: str, errors: list[dict[str, Any]]) -> dict[str, Any]:
    """
    Apply automatic fixes to project.

    This will:
    1. Create a backup of the project
    2. Apply suggested fixes
    3. Save the updated project
    4. Return the result

    Args:
        name: Project name
        errors: List of validation errors to fix

    Returns:
        Result of applying fixes including backup path
    """

    project_service: ProjectService = get_project_service()
    auto_fix_service = AutoFixService(project_service)

    validation_errors: list[ValidationError] = [ValidationError(**error) for error in errors]
    suggestions: list[FixSuggestion] = auto_fix_service.generate_fix_suggestions(validation_errors)
    result: FixResult = await auto_fix_service.apply_fixes(name, suggestions)

    if result.success:
        logger.info(f"Applied {result.fixes_applied} fixes to '{name}', backup at {result.backup_path}")
    else:
        logger.warning(f"Failed to apply fixes to '{name}': {result.errors}")

    return result.model_dump()

File: frontend/src/stores/validation.ts
import { defineStore } from 'pinia'
import { ref, computed } from 'vue'
import { api } from '@/api'
import type {
  ValidationResult,
  ValidationError,
  DependencyGraph,
  CircularDependencyCheck,
} from '@/types'

export const useValidationStore = defineStore('validation', () => {
  // State
  const validationResult = ref<ValidationResult | null>(null)
  const entityValidationResults = ref<Map<string, ValidationResult>>(new Map())
  const dependencyGraph = ref<DependencyGraph | null>(null)
  const circularDependencyCheck = ref<CircularDependencyCheck | null>(null)
  const loading = ref(false)
  const error = ref<string | null>(null)

  // Getters
  const hasErrors = computed(() => {
    return validationResult.value ? validationResult.value.error_count > 0 : false
  })

  const hasWarnings = computed(() => {
    return validationResult.value ? validationResult.value.warning_count > 0 : false
  })

  const errorCount = computed(() => validationResult.value?.error_count ?? 0)

  const warningCount = computed(() => validationResult.value?.warning_count ?? 0)

  const errors = computed(() => {
    return validationResult.value?.errors ?? []
  })

  const warnings = computed(() => {
    return validationResult.value?.warnings ?? []
  })

  const allMessages = computed(() => {
    return [...errors.value, ...warnings.value]
  })

  const messagesBySeverity = computed(() => {
    const grouped: Record<string, ValidationError[]> = {
      error: [],
      warning: [],
      info: [],
    }
    allMessages.value.forEach((msg: ValidationError) => {
      const severityGroup = grouped[msg.severity]
      if (severityGroup) {
        severityGroup.push(msg)
      }
    })
    return grouped
  })

  const messagesByEntity = computed(() => {
    const grouped: Record<string, ValidationError[]> = {}
    allMessages.value.forEach((msg: ValidationError) => {
      if (msg.entity) {
        if (!grouped[msg.entity]) {
          grouped[msg.entity] = []
        }
        const entityGroup = grouped[msg.entity]
        if (entityGroup) {
          entityGroup.push(msg)
        }
      }
    })
    return grouped
  })

  const hasCircularDependencies = computed(() => {
    return dependencyGraph.value?.has_cycles || circularDependencyCheck.value?.has_cycles || false
  })

  const cycles = computed(() => {
    return dependencyGraph.value?.cycles ?? circularDependencyCheck.value?.cycles ?? []
  })

  const cycleCount = computed(() => {
    return circularDependencyCheck.value?.cycle_count ?? cycles.value.length
  })

  const topologicalOrder = computed(() => {
    return dependencyGraph.value?.topological_order ?? []
  })

  const isValid = computed(() => {
    return validationResult.value?.is_valid ?? true
  })

  // Actions
  async function validateConfiguration(configName: string) {
    loading.value = true
    error.value = null
    try {
      validationResult.value = await api.projects.validate(configName)
      return validationResult.value
    } catch (err) {
      error.value = err instanceof Error ? err.message : 'Failed to validate configuration'
      throw err
    } finally {
      loading.value = false
    }
  }

  async function validateEntity(configName: string, entityName: string) {
    loading.value = true
    error.value = null
    try {
      const result = await api.validation.validateEntity(configName, entityName)
      entityValidationResults.value.set(entityName, result)
      return result
    } catch (err) {
      error.value = err instanceof Error ? err.message : 'Failed to validate entity'
      throw err
    } finally {
      loading.value = false
    }
  }

  async function fetchDependencies(configName: string) {
    loading.value = true
    error.value = null
    try {
      dependencyGraph.value = await api.validation.getDependencies(configName)
      return dependencyGraph.value
    } catch (err) {
      error.value = err instanceof Error ? err.message : 'Failed to fetch dependencies'
      throw err
    } finally {
      loading.value = false
    }
  }

  async function checkCircularDependencies(configName: string) {
    loading.value = true
    error.value = null
    try {
      circularDependencyCheck.value = await api.validation.checkCircularDependencies(configName)
      return circularDependencyCheck.value
    } catch (err) {
      error.value =
        err instanceof Error ? err.message : 'Failed to check circular dependencies'
      throw err
    } finally {
      loading.value = false
    }
  }

  function getEntityValidation(entityName: string): ValidationResult | null {
    return entityValidationResults.value.get(entityName) || null
  }

  function hasEntityErrors(entityName: string): boolean {
    const result = getEntityValidation(entityName)
    return result ? result.error_count > 0 : false
  }

  function hasEntityWarnings(entityName: string): boolean {
    const result = getEntityValidation(entityName)
    return result ? result.warning_count > 0 : false
  }

  function clearValidation() {
    validationResult.value = null
    entityValidationResults.value.clear()
  }

  function clearDependencies() {
    dependencyGraph.value = null
    circularDependencyCheck.value = null
  }

  function clearError() {
    error.value = null
  }

  function reset() {
    validationResult.value = null
    entityValidationResults.value.clear()
    dependencyGraph.value = null
    circularDependencyCheck.value = null
    loading.value = false
    error.value = null
  }

  return {
    // State
    validationResult,
    entityValidationResults,
    dependencyGraph,
    circularDependencyCheck,
    loading,
    error,
    // Getters
    hasErrors,
    hasWarnings,
    errorCount,
    warningCount,
    errors,
    warnings,
    allMessages,
    messagesBySeverity,
    messagesByEntity,
    hasCircularDependencies,
    cycles,
    cycleCount,
    topologicalOrder,
    isValid,
    // Actions
    validateConfiguration,
    validateEntity,
    fetchDependencies,
    checkCircularDependencies,
    getEntityValidation,
    hasEntityErrors,
    hasEntityWarnings,
    clearValidation,
    clearDependencies,
    clearError,
    reset,
  }
})
